### PR TITLE
Load responses by ajax by default on non-question views of instructor results #5417

### LINKED ISSUE #5417: Load responses by ajax by default on non-question views of instructor results
The questions view of the instructors results page automatically load responses incrementally by ajax. This solved some problems with a slow initial load of the results page. Currently for the non-question views, we try to render all responses at once unless there are a large number of responses. We can make the non-question views, such as recipient > giver > question, do the same and use ajax loading by default. This will help in making the code more understandable on the results page too, as there will be less possible ways that the results page can loaded.

### PR DESCRIPTION
**Pull request by belyabl9**
Fixes #5417 

This is the initial version. There are lots of static .html changes, so, could somebody more experienced with that data samples check that I missed nothing ?

### CODE REVIEW

**Comment by kanghj** [comment]
@belyabl9, I'll try to finish reviewing this by tomorrow.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 1]
```diff
@@ -226,266 +226,189 @@ public void initForSectionPanelViews(InstructorAttributes instructor,
...
-            
+
             String currentTeam = getCurrentTeam(bundle, primaryParticipantIdentifier);
-            String currentSection = getCurrentSection(primaryToSecondaryParticipantToResponsesMap, viewType);
-            
+
+            boolean isStudent = bundle.isParticipantIdentifierStudent(primaryParticipantIdentifier);
+            String participantSection = bundle.getSectionFromRoster(primaryParticipantIdentifier);
+
+            if (isStudent && !participantSection.equals(section)) {
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: This is just for optimisation right? We can remove this line, we can improve the methods performance in a future issue

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 2]
```diff
@@ -226,266 +226,189 @@ public void initForSectionPanelViews(InstructorAttributes instructor,
...
+                     primaryToSecondaryParticipantToResponsesMap
+                : sortedResponses.entrySet()) {
             primaryParticipantIndex += 1;
             String primaryParticipantIdentifier = primaryToSecondaryParticipantToResponsesMap.getKey();
-            
+
+            boolean isStudent = bundle.isParticipantIdentifierStudent(primaryParticipantIdentifier);
+            String participantSection = bundle.getSectionFromRoster(primaryParticipantIdentifier);
+
+            if (isStudent && !participantSection.equals(section)) {
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Same for this. Let's remove this

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 3]
```diff
@@ -226,266 +226,189 @@ public void initForSectionPanelViews(InstructorAttributes instructor,
...
-                    participantsWithResponses.clear();
+                            sectionPanel, prevTeam, teamMembersWithResponses);
+                    teamMembersWithResponses.clear();
                 }
-                
+
                 teamsWithResponses.add(currentTeam);
-            }
-            
-            if (isDifferentSection) {
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: It's great that we can remove this :D Good work

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 4]
```diff
@@ -226,266 +226,189 @@ public void initForSectionPanelViews(InstructorAttributes instructor,
...
-        }
-        
+        buildMissingParticipantPanelsForTeam(sectionPanel, prevTeam, teamMembersWithResponses);
+
         teamsWithResponses.add(prevTeam);
-        if (isGroupedByTeam()) {
-            buildMissingTeamAndParticipantPanelsForSection(
-                                            sectionPanel, prevSection, teamsWithResponses);
-        } else {
-            buildMissingParticipantPanelsForSection(sectionPanel, prevSection, participantsWithResponses);
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: When not grouped by teams, the missing participants are incorrectly generated. This can be reproduced when the students are not ordered in the teams alphabetically.

Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Perhaps our test cases do not cover this case well, I'll check and make an issue to change the tests in a separate issue.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 2 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Let's remove this

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 5]
```diff
@@ -862,60 +720,34 @@ private void buildMissingParticipantPanelsForTeam(
...
-                if (viewType.isPrimaryGroupingOfGiverType()) {
-                    addMissingParticipantsPanelsWithModerationButtonForTeam(
-                                                    sectionPanel, teamInMissingSection, teamMembers);
-                } else {
-                    addMissingParticipantsPanelsWithoutModerationButtonForTeam(
-                                                    sectionPanel, teamInMissingSection, teamMembers);
-                }
-            }
-        }
+    private InstructorFeedbackResultsSectionPanel buildSectionPanel(String name, String nameForDisplay, boolean loadByAjax) {
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Instead of this, let's either have a constructor or factory method for `InstructorFeedbackResultsSectionPanel ` that accepts these 3 parameters.

File: src/main/webapp/WEB-INF/tags/instructor/results/sectionPanel.tag [hunk 6]
```diff
@@ -36,20 +36,18 @@
             </div>
         </c:when>
         <c:otherwise>
-            <div class="panel-heading ajax_submit">
+            <div class="panel-heading ajax_submit ajax_auto">
```
Comment (kanghj) [inline src/main/webapp/WEB-INF/tags/instructor/results/sectionPanel.tag]: I think we can remove `ajax_submit`, since all panels are always automatically loaded now.

File: src/main/webapp/js/instructorFeedbackResultsAjaxByGQR.js [hunk 7]
```diff
@@ -28,35 +27,44 @@ $(document).ready(function() {
                 displayAjaxRetryMessageForPanelHeading(displayIcon);
             },
             success: function(data) {
+                var sectionBody = panelBody[0];
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackResultsAjaxByGQR.js]: A minor point, but let's improve this a step further by directly `var $sectionBody = $(panelBody[0])`

File: src/main/webapp/js/instructorFeedbackResultsAjaxByGQR.js [hunk 8]
```diff
@@ -1,15 +1,14 @@
 var numPanels = 0;
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackResultsAjaxByGQR.js]: Are we able to move this declaration to line 5?

File: src/main/webapp/js/instructorFeedbackResultsAjaxByGRQ.js [hunk 9]
```diff
@@ -28,39 +26,43 @@ $(document).ready(function() {
...
-                    $(panelBody[0]).html('The results is too large to be viewed. Please choose to view '
-                                         + 'the results by questions or download the results.');
+                    $(sectionBody).html('The results is too large to be viewed. '
+                                      + 'Please choose to view the results by questions or download the results.');
                 } else {
                     var appendedSection = $(data).find('#sectionBody-0').html();
+                    var panelsInSection = $(appendedSection).find('div.panel');
+
+                    if (panelsInSection.length === 0) {
+                        $(panelHeading).parent().remove();
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackResultsAjaxByGRQ.js]: Let's add a comment on what this is doing or extract this to a new function

File: src/main/webapp/jsp/instructorFeedbackResultsByRecipientGiverQuestion.jsp [hunk 10]
```diff
@@ -4,7 +4,7 @@
 <%@ taglib tagdir="/WEB-INF/tags/instructor/results" prefix="results" %>
 <c:set var="jsIncludes">
     <script type="text/javascript" src="/js/feedbackResponseComments.js"></script>
-    <script type="text/javascript" src="/js/instructorFeedbackResultsAjaxByRQG.js"></script>
+    <script type="text/javascript" src="/js/instructorFeedbackResultsAjaxByRGQ.js"></script>
```
Comment (kanghj) [inline src/main/webapp/jsp/instructorFeedbackResultsByRecipientGiverQuestion.jsp]: Good catch, thanks for fixing this

Comment (kanghj) [inline src/main/webapp/jsp/instructorFeedbackResultsByRecipientGiverQuestion.jsp]: Hopefully now that we are minimising the number of different ways the page can load, this sort of bugs can be detected more quickly.

**Comment by kanghj** [comment]
It seems the behaviour for adding FeedbackResponseComments is wrong now when trying to add response comments for sections which are not the first section. It incorrectly opens the form for adding response comments to a participant in the first section instead.

File: src/main/webapp/js/instructorFeedbackResults.js [hunk 11]
```diff
@@ -211,10 +211,28 @@ function bindCollapseEvents(panels, nPanels) {
...
-            $(heading[0]).attr('data-target', '#panelBodyCollapse-' + numPanels);
-            $(heading[0]).attr('id', 'panelHeading-' + numPanels);
+
+            var sectionIndex = '';
+
+            var isSectionPanel = true;
+            var sectionBody = $(heading).next().find('[id^="sectionBody-"]');
+
+            if (sectionBody.length === 0) {
+                sectionBody = $(heading).parents('[id^="sectionBody-"]');
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackResults.js]: When is this branch taken? Might need to add a comment or an explaining variable here

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 12]
```diff
@@ -247,8 +247,13 @@ public void waitForPanelsToCollapse() {
     public void waitForPanelsToExpand() {
         By panelCollapseSelector = By.cssSelector(".panel-heading+.panel-collapse");
         List<WebElement> webElements = browser.driver.findElements(panelCollapseSelector);
-        
-        waitForElementsVisibility(webElements);
+        for (WebElement element : webElements) {
+            try {
+                waitForElementVisibility(element);
+            } catch (org.openqa.selenium.StaleElementReferenceException e) {
```
Comment (kanghj) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: Let's import `StaleElementReferenceException` and use it directly without its fully qualified name

**Comment by kanghj** [comment]
Some of the testing html changes looks a bit unexpected to me, I'll have a closer look again later.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 1 above)
Comment (belyabl9) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: No, as I understand getCurrentSection() determines student's section by his first response. I was confused about situation when student's section was changed but he is still shown as a main recipient for other section. So, I changed the way to determine the current section

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 2 above)
Comment (belyabl9) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Answered in previous comment

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 1 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: I don't quite understand the situation you are describing. How do I reproduce that bug? Is this related to the problem we have when a student is also an instructor?

Comment (belyabl9) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: When I tested these changes I spent a lot of time until I realize the source of problem
Example:
"Student A" from "Section 1" has left a feedback response in session A
Then, "Student A" has been moved to "Section 2" 
getCurrentSection() determines student's section via his first found response
So, in "Recipient -> * -> *" view this student will be shown in "Section 1" that seems to be wrong as for me as it is not actual information and may confuse users

Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Ah, got it. #4486 was supposed to stop this sort of errors from happening, but that didn't seem to have gotten anywhere.

**Comment by kanghj** [comment]
A possibly unintended consequence is that sections without any responses are not shown anymore, whereas beforehand, the section panel would list the possible/missing participants. We should be able to live with this, I guess, since the "Participants who have not responded to any question" already shows the missing participants anyway.

**Comment by kanghj** [comment]
Most of the differences in html seem to be coming from the now incorrect 'missing participants' panels, and that sections without any responses are not shown. So let's fix the missing participants panels first before doing a next round of review.

**Comment by belyabl9** [comment]
@kanghj , what about sections without responses, I did it intentionally. I tried to repeat behavior on master and I saw that 'Not in a section'  is hidden when there are no responses there.
So, let's clarify the expected behavior. Is the following correct:
    1. If section is 'Not in a section' without responses we hide it
    2. If any other section doesn't have responses we build it with missing participant panels
?
By the way, thanks for such detailed review. I'm going to make changes in the nearest time

**Comment by kanghj** [comment]
> 1. If section is 'Not in a section' without responses we hide it
2. If any other section doesn't have responses we build it with missing participant panels

@belyabl9, yes that's currently the expected behavior on master.

**Comment by kanghj** [comment]
@belyabl9, how is this coming along? If the complexity of this PR grows too much, remember you can do smaller parts of this in separate issues.

**Comment by belyabl9** [comment]
@kanghj , I'm going to finish it within this PR. As I see, there is one problem has left - wrong missing participants when not grouped by teams. I pushed partial commits just to sync with my laptop. I'll mention when it is ready to review

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 4 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: For future developers' reference, here are the settings to reproduce the incorrect missing participants, as discussed:

class_roster
(The demo course students, but renamed "Alice" to "Jenny", and added an additional team whose student "Jean" < "Jenny".)

Use a non-question, but without grouping by team:
results_filter

Delete all responses from emma: this causes the "missing participants" logic to generate missing responses for emma' entire team
incorrect_missing_participant

Expected result:
master

@belyabl9 sorry I have not yet looked through the existing test cases as promised. I will try to do it soon.

**Comment by belyabl9** [comment]
@kanghj , I've done the requested changes. What about the missing participants, I did it exactly as you show, but on the latest code I couldn't achieve that wrong results. If it's not difficult, please, re-check it on current version or I can wait for the test case. This issue is getting huge and it becomes difficult to rebase it, so it would be great to pass this issue if it doesn't cause any critical problems and fix/refactor small issues separately

**Comment by kanghj** [comment]
@belyabl9, for the missing participants, a place where it is more obvious to see that the behavior is wrong is in InstructorFeedbackResultsPage [line 217](https://github.com/TEAMMATES/teammates/blob/master/src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java#L217): `resultsPage.verifyHtmlMainContent("/instructorFeedbackResultsSortGiverRecipientQuestionTeam.html");`. I added a student in another pull request a few days back to make this more clear. Can you have a look at that test case? The part of the code that I commented in the diff should be related to this behavior. 

The pull request's increasing size and complexity is precisely the reason why we should split the pull request up into separate smaller pull requests. Ideally, each PR should at be most 2 work days (e.16) worth of work, for ease of syncing with master and reviewing. In future, try to avoid enormous pull requests like this.

File: src/main/webapp/js/feedbackResponseComments.js [hunk 13]
```diff
@@ -304,8 +330,10 @@ function showResponseCommentAddForm(recipientIndex, giverIndex, qnIndx) {
     $('#responseCommentAddForm' + id).focus();
 }
 
-function hideResponseCommentAddForm(recipientIndex, giverIndex, qnIndx) {
-    var id = '-' + recipientIndex + '-' + giverIndex + '-' + qnIndx;
+function hideResponseCommentAddForm(/* firstIndex, secondIndex, thirdIndex, fourthIndex */) {
+    var isIncludeSections = arguments.length === 4;
```
Comment (kanghj) [inline src/main/webapp/js/feedbackResponseComments.js]: I would prefer if the arguments are named. This way of accessing function parameters will make it very difficult for any future maintainer. For the `sectionIndex`, pass an object `opt` containing `sectionIndex` as a field. We should do it in this PR because doing this requires context of what the old arguments were, and a future developer taking this as a new issue would not have knowledge of this.

File: src/main/webapp/js/feedbackResponseComments.js [hunk 14]
```diff
@@ -314,13 +342,44 @@ function hideResponseCommentAddForm(recipientIndex, giverIndex, qnIndx) {
...
+        }
     } else {
         id = '-' + recipientIndex;
     }
+
+    return id;
+}
+
+function showResponseCommentEditForm(/* firstIndex, secondIndex, thirdIndex, fourthIndex, fifthIndex */) {
+    var isIncludeSections = arguments.length === 5;
```
Comment (kanghj) [inline src/main/webapp/js/feedbackResponseComments.js]: Same for this. I would prefer if the arguments are named. This way of accessing function parameters will make it very difficult for any future maintainer. For the sectionIndex, pass an object opt containing sectionIndex as a field. We should do it in this PR because doing this requires context of what the old arguments were, and a future developer taking this as a new issue would not have knowledge of this.

File: src/main/webapp/js/feedbackResponseComments.js [hunk 15]
```diff
@@ -329,44 +388,75 @@ function showResponseCommentEditForm(recipientIndex, giverIndex, qnIndex, commen
...
     } else {
-        id = '-' + sessionIdx;
+        id = '-' + sessionIndex + '-' + questionIndex + '-' + responseIndex;
     }
+
+    return id;
+}
+
+function toggleVisibilityEditForm(/* firstIndex, secondIndex, thirdIndex, fourthIndex, fifthIndex */) {
+    var isIncludeSections = arguments.length === 5;
```
Comment (kanghj) [inline src/main/webapp/js/feedbackResponseComments.js]: Same for this. I would prefer if the arguments are named. This way of accessing function parameters will make it very difficult for any future maintainer. For the sectionIndex, pass an object opt containing sectionIndex as a field. We should do it in this PR because doing this requires context of what the old arguments were, and a future developer taking this as a new issue would not have knowledge of this.

File: src/main/webapp/js/instructorFeedbackResults.js [hunk 16]
```diff
@@ -231,6 +256,24 @@ function displayAjaxRetryMessageForPanelHeading($element) {
...
+ */
+function removeEmptySection(id, content) {
+    var $heading = $('[id^=panelHeading-section-' + id + ']');
+
+    if (parseInt(id, 10) === 0) {
+        var panelsInSection = content.find('div.panel');
+
+        if (panelsInSection.length === 0) {
+            $heading.parent().remove();
+            return 1;
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackResults.js]: Looks like this function should be returning booleans

File: src/main/webapp/js/instructorFeedbackResults.js [hunk 17]
```diff
@@ -231,6 +256,24 @@ function displayAjaxRetryMessageForPanelHeading($element) {
...
+        }
+    }
+
+    return 0;
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackResults.js]: Same here. Use `false` instead of 0

File: src/main/webapp/js/feedbackResponseComments.js [hunk 18]
```diff
@@ -329,44 +388,75 @@ function showResponseCommentEditForm(recipientIndex, giverIndex, qnIndex, commen
...
             .html('<span class="glyphicon glyphicon-eye-close"></span> Show Visibility Options');
-        
     } else {
         visibilityEditForm.show();
         $('#frComment-visibility-options-trigger' + id)
             .html('<span class="glyphicon glyphicon-eye-close"></span> Hide Visibility Options');
     }
 }
 
-function hideResponseCommentEditForm(recipientIndex, giverIndex, qnIndex, commentIndex) {
```
Comment (kanghj) [inline src/main/webapp/js/feedbackResponseComments.js]: What happened to this?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 19]
```diff
@@ -614,11 +537,18 @@ private InstructorFeedbackResultsGroupByParticipantPanel buildGroupByParticipant
                                                         response.feedbackSessionName,
                                                         Const.ParamsNames.INSTRUCTOR_PERMISSION_SUBMIT_SESSION_IN_SECTIONS);
             
+            int sectionId = 0;
+            Pattern pattern = Pattern.compile("^section-(\\d+)");
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Let's move this to be a static field, we don't have to initialise this each time

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 1 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Which test/sample feedback session were you observing this in? I think we should remove this check, this is the concern of #4486, not here.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java [hunk 20]
```diff
@@ -124,7 +124,9 @@ protected ActionResult execute() throws EntityDoesNotExistException {
         data.commentId = commentId;
         data.showCommentToString = joinParticipantTypes(createdComment.showCommentTo, ",");
         data.showGiverNameToString = joinParticipantTypes(createdComment.showGiverNameTo, ",");
-        
+
+        data.setCommentIds(commentId);
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java]: Can you explain why this needs to be done? There are a few places where `InstructorFeedbackResponseCommentAjaxPageData` is used, but this is not done in the other places.

**Comment by kanghj** [comment]
Some clarifications requested.

File: src/main/webapp/js/instructorFeedbackResultsAjaxByGQR.js [hunk 21]
```diff
@@ -28,35 +24,40 @@ $(document).ready(function() {
...
-                    $(panelBody[0]).html('The results is too large to be viewed. Please choose to view '
-                                         + 'the results by questions or download the results.');
+                    $sectionBody.html('The results is too large to be viewed. '
+                            + 'Please choose to view the results by questions or download the results.');
                 } else {
-                    var appendedSection = $(data).find('#sectionBody-0').html();
+                    var $appendedSection = $(data).find('#sectionBody-0');
+                    var sectionId = $(panelHeading).attr('id').match(/section-(\d+)/)[1];
+
+                    if (removeEmptySection(sectionId, $appendedSection)) {
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackResultsAjaxByGQR.js]: `removeEmptySection` is doing too much, let's split it up into `isEmptySection` and `removeSection`.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java (hunk 20 above)
Comment (belyabl9) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java]: Current implementation initializes commentIds only while calling getComment() method, so I needed a way to initialize it before passing to .jsp

**Comment by belyabl9** [comment]
@kanghj , you were right about wrong missing participant generation. As I see enabling sorting by group for all the cases (not only when check box is checked) solves this problem
Other requested changes have been done as well
Ready to review

**Comment by kanghj** [comment]
IMO, fixing it by always sorting the recipients by team name sort of defeats the purpose of the "grouped by team" option/checkbox. In the old behavior, when that option is not selected, the participants are sorted alphabetically by name, whereas now they are still sorted by team first, then name. But perhaps this change in behavior is fine for now, it does not seem very important and we can fix in another issue. @damithc, what do you think?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 1 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: @belyabl9, let's have some clarification on this. I will like to fix this separately if possible.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 22]
```diff
@@ -613,12 +538,18 @@ private InstructorFeedbackResultsGroupByParticipantPanel buildGroupByParticipant
                     && instructor.isAllowedForPrivilege(response.recipientSection,
                                                         response.feedbackSessionName,
                                                         Const.ParamsNames.INSTRUCTOR_PERMISSION_SUBMIT_SESSION_IN_SECTIONS);
-            
+
+            Pattern pattern = Pattern.compile("^section-(\\d+)");
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Same point as the previous round of review, `Pattern` can be made into a static field?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java (hunk 20 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java]: Let's fix this seperately in #5689 first

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 22 above)
Comment (belyabl9) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Sure, I thought you meant just sectionId

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentAdd.tag [hunk 23]
```diff
@@ -7,8 +7,19 @@
...
-<c:set var="divId" value="${firstIndex}-${secondIndex}-${thirdIndex}" />
-<c:set var="divIdAsJsParams" value="${firstIndex},${secondIndex},${thirdIndex}" />
+<%@ attribute name="fourthIndex" %>
+
+<c:choose>
+    <c:when test="${not empty firstIndex && not empty secondIndex && not empty thirdIndex && not empty fourthIndex}">
+        <c:set var="divId" value="${fourthIndex}-${firstIndex}-${secondIndex}-${thirdIndex}" />
+        <c:set var="divIdAsJsParams" value="${firstIndex},${secondIndex},${thirdIndex}, { sectionIndex: ${fourthIndex} }" />
+    </c:when>
+    <c:when test="${not empty firstIndex && not empty secondIndex && not empty thirdIndex}">
```
Comment (kanghj) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentAdd.tag]: This seems a bit more complicated than it should be. I might be missing something, but are we able to simplify this entire `c:choose` to a check if `fourthIndex` is not empty?

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag [hunk 24]
```diff
@@ -25,7 +25,13 @@
...
         <a id="frComment-visibility-options-trigger-${divId}"
            class="btn btn-sm btn-info pull-right"
-           onclick="toggleVisibilityEditForm(${divIdAsJsParams})">
+            <c:if test="${isEditForm}">
+                onclick="toggleVisibilityEditForm(${divIdAsJsParams})">
+            </c:if>
+            <c:if test="${isAddForm}">
+                onclick="toggleVisibilityAddForm(${divIdAsJsParams})">
+            </c:if>
+        >
```
Comment (kanghj) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: Seems to be a stray `>` here

Comment (belyabl9) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: I'm not sure I got it. Do you mean </c:if>**>** ?

Comment (kanghj) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: I mean the `>` on line 34, the `<a>` tag is closed on either line 29 or line 32 (`onclick="toggle...">`) already. The html changes on `instructorCommentsPageAddFrc.html` contains a new `&gt;`.

Comment (chowyb) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: The `>` here appears to close the `a` tag starting at line 26

Comment (belyabl9) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: Oh, I see, thanks

**Comment by belyabl9** [comment]
Ready to review

File: src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java (hunk 20 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResponseCommentAddAction.java]: @belyabl9, this should already be fixed in #5689, we can remove this. Not sure why the changes there did not show up here.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 25]
```diff
@@ -47,6 +49,9 @@
     
     private static final int RESPONDENTS_LIMIT_FOR_AUTOLOADING = 150;
 
+    private static int sectionId;
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Opps, this probably shouldn't be a field. I meant the `Pattern` (which you have now done) in that comment a few days ago, so this sectionId can remain a local variable where it is used

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 1 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: @belyabl9, ping on this

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentAdd.tag [hunk 26]
```diff
@@ -7,8 +7,19 @@
...
-<c:set var="divId" value="${firstIndex}-${secondIndex}-${thirdIndex}" />
-<c:set var="divIdAsJsParams" value="${firstIndex},${secondIndex},${thirdIndex}" />
+<%@ attribute name="fourthIndex" %>
+
+<c:choose>
+    <c:when test="${not empty fourthIndex}">
+        <c:set var="divId" value="${fourthIndex}-${firstIndex}-${secondIndex}-${thirdIndex}" />
+        <c:set var="divIdAsJsParams" value="${firstIndex},${secondIndex},${thirdIndex}, { sectionIndex: ${fourthIndex} }" />
+    </c:when>
+    <c:when test="${not empty firstIndex && not empty secondIndex && not empty thirdIndex}">
```
Comment (kanghj) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentAdd.tag]: Hmm, should this be a `<c:otherwise>` instead?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java (hunk 1 above)
Comment (belyabl9) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: I'll remove this check from this PR

**Comment by belyabl9** [comment]
@kanghj , please, look at changes

**Comment by kanghj** [comment]
The behavior to hide the "Not in a section" panel only after ajax may get annoying in large sessions with many responses. But I guess we can fix that in another issue, since this PR is getting really large. 

Similarly, if there are many students in the course, we should try to fallback on the old behavior where the instructor needs to click on each section panel to expand it. But again, this should be done in another issue.

When not grouped by team, the "collapse/expand students" button does not work on 'missing' participants. This is the fault of the existing code handling ajax loading of the page though. But try to see if it is possible to fix this easily, @belyabl9? There is code duplication between `instructorFeedbackResults.js` (around line 274) and the different js files for handling ajax, which is the cause of this problem. If it can't be fixed easily in this issue, let's make a new one to fix it since it already currently affects the results page on non-question views when loaded by ajax.

**Comment by belyabl9** [comment]
@kanghj , I think mentioned issues and refactorings should be done in other issues in order not to waste a lot of time maintaining all the current code in this issue, rebasing it and so on.

**Comment by kanghj** [comment]
Ok,  let's try that then

**Comment by belyabl9** [comment]
Rebased

**Comment by belyabl9** [comment]
Rebased
